### PR TITLE
Sqlite improvements

### PR DESCRIPTION
**Pull request by bhansconnect**
I think this will be a really cool api improvement that also should give us some perf gains. That said, currently breaks the compiler. Putting the PR up so that other can look at it. Maybe someone else has some ideas of things to mess with to get it compiling. My guess is that tasks + record builders do not play happily together, but the root cause my be something else.

Really want to land this if possible, but might take compiler fixes. I think the impl is pretty cool. Thanks @agu-z for the record bulder help.

### CODE REVIEW

**Comment by bhansconnect** [comment]
Ok, re-aimed this at main now that refactor host is merged.

**Comment by bhansconnect** [comment]
Update here. Have been working with Luke and iterating on both api and usability. So Things are shifting again now.

**Comment by bhansconnect** [comment]
@Anton-4, I renamed some files and am getting CI failures now. Probably am missing something obvious, but I can't seem to figure out why it is still pulling in the old name.

**Comment by Anton-4** [comment]
I fixed the old InternalSql references, perhaps you forgot to commit a file?

Now, they all get stuck at:
```
 + roc build --prebuilt-platform --linker=legacy ./examples/sqlite.roc
thread '<unnamed>' panicked at crates/compiler/mono/src/ir.rs:6132:56:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by bhansconnect** [comment]
> I fixed the old InternalSql references, perhaps you forgot to commit a file?

Yeahs, something must not be matching locally with what I have pushed. I know I updated those file...

>  + roc build --prebuilt-platform --linker=legacy ./examples/sqlite.roc
thread '<unnamed>' panicked at crates/compiler/mono/src/ir.rs:6132:56:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

Yeah, this broke recently when I was trying to make the tags nicer. Have to figure it out still.

**Comment by bhansconnect** [comment]
For the sql file. Locally, I had `InternalSql.roc`, but apparently my git isn't case sensitive? I had to full delete the file and re-add it to get git to view rename the file to `InternalSql.roc`.

**Comment by Anton-4** [comment]
> but apparently my git isn't case sensitive?

Huh, that's unexpected :smile:

**Comment by lukewilliamboswell** [comment]
```
$ roc run examples/todos.roc
🔨 Rebuilding platform...
thread 'main' panicked at crates/compiler/gen_llvm/src/llvm/build.rs:5770:19:
Error in alias analysis: error in module ModName("UserApp"), function definition FuncName("(\x00\x00\x00^\x00\x00\x00\x18:;\x90\xb2R\x9d?"): expected type 'union { (union { ((),), ((heap_cell,),), (), ((heap_cell,), union { (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), () }), (), ((),), () },), ((heap_cell, bag<(heap_cell,)>),) }', found type 'union { (union { ((),), ((heap_cell,),), ((heap_cell,), union { (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), (), () }), ((),) },), ((heap_cell, bag<(heap_cell,)>),) }'
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: crates/roc_host/src/roc.rs [hunk 1]
```diff
@@ -771,170 +785,313 @@ fn os_str_to_roc_path(os_str: &std::ffi::OsStr) -> RocList<u8> {
...
+    // Always prepare once to ensure no errors and prep for current thread.
+    if let Err(err) = thread_local_prepare(&stmt) {
+        return roc_err_from_sqlite_err(err);
+    }
+
+    let heap = sqlite_stmt_heap();
+    let alloc_result = heap.alloc_for(stmt);
+    match alloc_result {
+        Ok(out) => {
+            // TODO: only make these constant if they escape init.
```
Comment (bhansconnect) [inline crates/roc_host/src/roc.rs:909]: I do think this TODO needs to be dealt with before we land. Just an FYI

Comment (lukewilliamboswell) [reply crates/roc_host/src/roc.rs]: @bhansconnect  I think I resolved this in fd317fb. Can you just double check as I wasn't 100% around the logic here. Like are we setting refcount to zero during or after init. I've opted for during as that feels right, these stmts will escape init and therefore we don't want them to be freed.

Comment (bhansconnect) [reply crates/roc_host/src/roc.rs]: I think I realized a cooler way to do this. Just pushed it. It also has the advantage that an sqlite statement that is freed during init will actually get freed. Only those still live after init will be promoted to constants!

**Comment by bhansconnect** [comment]
Ok, I think this is good for final review and merge!

**Review by lukewilliamboswell — APPROVED** [review]
Woo! 🐙

**Pull request merged**